service/src/main/java/bio/terra/workspace/service/resource/referenced/flight/clone/CloneReferenceResourceStep.java

```diff
@@ -0,0 +1,106 @@
+import bio.terra.stairway.StepResult;
+import bio.terra.stairway.exception.RetryException;
+import bio.terra.workspace.common.utils.FlightUtils;
+import bio.terra.workspace.service.iam.AuthenticatedUserRequest;
+import bio.terra.workspace.service.resource.model.StewardshipType;
+import bio.terra.workspace.service.resource.referenced.ReferencedResourceService;
+import bio.terra.workspace.service.resource.referenced.model.ReferencedResource;
+import bio.terra.workspace.service.workspace.flight.WorkspaceFlightMapKeys.ControlledResourceKeys;
+import bio.terra.workspace.service.workspace.model.WsmCloneResourceResult;
+import bio.terra.workspace.service.workspace.model.WsmResourceCloneDetails;
+import com.fasterxml.jackson.core.type.TypeReference;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import java.util.UUID;
+import org.springframework.http.HttpStatus;
+
+public class CloneReferenceResourceStep implements Step {
+
+  private final AuthenticatedUserRequest userRequest;
+  private final ReferencedResourceService referencedResourceService;
+  private final ReferencedResource sourceResource;
+  private final ReferencedResource destinationResource;
+
+  public CloneReferenceResourceStep(
+      AuthenticatedUserRequest userRequest,
+      ReferencedResourceService referencedResourceService,
+      ReferencedResource sourceResource,
+      ReferencedResource destinationResource) {
+    this.userRequest = userRequest;
+    this.referencedResourceService = referencedResourceService;
+    this.sourceResource = sourceResource;
+    this.destinationResource = destinationResource;
+  }
+
+  @Override
+  public StepResult doStep(FlightContext context) throws InterruptedException, RetryException {
+    WsmResourceCloneDetails cloneDetails =
+        new WsmResourceCloneDetails()
+            .setStewardshipType(StewardshipType.REFERENCED)
+            .setResourceType(sourceResource.getResourceType())
+            .setSourceResourceId(destinationResource.getResourceId())
+            .setName(destinationResource.getName())
+            .setDescription(destinationResource.getDescription());
+
+    cloneDetails
+        .setResourceType(destinationResource.getResourceType())
+        .setStewardshipType(destinationResource.getStewardshipType())
+        .setCloningInstructions(destinationResource.getCloningInstructions())
+        .setDestinationResourceId(destinationResource.getResourceId());
+
+    try {
+      var createdResource =
+          referencedResourceService.createReferenceResourceForClone(
+              destinationResource, sourceResource, userRequest);
+      FlightUtils.setResponse(context, createdResource, HttpStatus.OK);
+      cloneDetails.setResult(WsmCloneResourceResult.SUCCEEDED);
+    } catch (Exception e) {
+      cloneDetails.setResult(WsmCloneResourceResult.FAILED).setErrorMessage(e.getMessage());
+    }
+
+    // add to the map
+    addCloneDetailsToWorkingMap(context, cloneDetails);
```

Comment:
The error case appears to fall through. Is that what you intend?
Maybe some comment about what is going on?

Reply:
it should fail here. Fixed.